lib/queuePopulator/KafkaLogConsumer/LogConsumer.js

```diff
@@ -42,70 +46,141 @@ class LogConsumer {
+            method: 'LogConsumer._onOffsetCommit',
+            topicPartitions,
+            groupId: this._groupId
+        });
+        return undefined;
     }
 
     /**
-     * Queries last commited partition offsets
-     * and stores them
-     * @param {Function} cb callback
-     * @returns {undefined}
+     * @param {kafka.KafkaError} err Rebalance event
+     * @param {TopicPartition[]} assignment List of (un)assigned partitions
+     * @returns {void}
      */
-    _storeCurrentOffsets(cb) {
-        this._consumer.committed(5000, (err, topicPartitions) => {
-            if (err) {
-                this._log.error('Error while getting offsets', {
-                    method: 'LogConsumer._storeCurrentOffsets',
+    _onRebalance(err, assignment) {
+        if (err.code === kafka.CODES.ERRORS.ERR__ASSIGN_PARTITIONS) {
+            this._log.info('rdkafka.assign', {
+                method: 'LogConsumer._onRebalance',
+                assignment,
+                topic: this._topic,
+                consumerGroupId: this._consumerGroupId,
+            });
+            try {
+                this._consumer.assign(assignment);
+            } catch (e) {
+                const logger = this._consumer.isConnected() ? this._log.error : this._log.debug;
+                logger.bind(this._log)('rdkafka.assign failed', {
+                    method: 'LogConsumer._onRebalance',
+                    error: e.toString(),
+                    assignment,
                     topic: this._topic,
-                    error: err.message,
                     consumerGroupId: this._consumerGroupId,
                 });
-                return cb();
             }
-            // saving offsets
-            this._offsets = topicPartitions;
-            return cb();
-        });
+        } else if (err.code === kafka.CODES.ERRORS.ERR__REVOKE_PARTITIONS) {
+            this._log.info('rdkafka.revoke', {
+                method: 'LogConsumer._onRebalance',
+                assignment,
+                topic: this._topic,
+                consumerGroupId: this._consumerGroupId,
+            });
+            const startTime = Date.now();
+            const rebalanceTimeoutInterval = setInterval(() => {
+                const elapsed = Date.now() - startTime;
+                if (elapsed >= this._maxPollIntervalMs - 1000) {
+                    this._log.error('rdkafka.unassign timeout: consumer stuck, disconnecting', {
+                        method: 'LogConsumer._onRebalance',
+                        topic: this._topic,
+                        consumerGroupId: this._consumerGroupId,
+                    });
+                    clearInterval(rebalanceTimeoutInterval);
+                    this._consumer.disconnect();
+                    return;
+                }
+                // wait for consumed messages to finish being processed
+                if (this._topicPartition && this._topicPartition.length > 0) {
```

Comment:
instead of polling, can't we trigger this callback when the topic gets empty?
similar to BackbeatConsumer, we could create a `drain` callback which is called when `_topicPartition` is reset ; and set it when needed in rebalance.